### PR TITLE
examples/block-ip: show how to block IP addresses

### PR DESCRIPTION
**Pull request by jay**
- Show how CURLOPT_OPENSOCKETFUNCTION can be used to block IP addresses.

This is a new example that demonstrates IP blocking.

Ref: https://github.com/curl/curl/discussions/15710#discussioncomment-11534877

Closes #xxxx

### CODE REVIEW

**Comment by jay** [comment]
Is `-Werror=missing-prototypes` really needed for examples?

**Comment by vszakats** [comment]
> Is `-Werror=missing-prototypes` really needed for examples?

It adds complexity to apply and remember different rules for
different parts of the codebase. Examples can be copied and
compiled by users outside of curl and may have this warning
enabled there.

Isn't it simpler to add `static` to the 3 local functions?

Silencing the struct warning should also be something done in
other parts of the code.

**Comment by bagder** [comment]
> Isn't it simpler to add `static` to the 4 local functions?

I believe so. We encourage use of static for local functions anyway.

**Review by bagder — APPROVED** [review]

**Review by vszakats — APPROVED** [review]

**Comment by jay** [comment]
@vszakats do you have any suggestion for the failing amiga CI job? Does it not have a working inet ntop/pton? I see in your commit that recently created the job that: *inet_ntop, inet_pton: fix using it for AmigaOS. cmake detects them, and they did not compile with AmigaOS*. Should this example be excluded from that platform?

**Comment by vszakats** [comment]
> @vszakats do you have any suggestion for the failing amiga CI job? Does it not have a working inet ntop/pton? I see in your commit that recently created the job that: _inet_ntop, inet_pton: fix using it for AmigaOS. cmake detects them, and they did not compile with AmigaOS_. Should this example be excluded from that platform?

These functions exist on AmigaOS3, but they possibly require `proto/bsdsocket.h` (can't double-check ATM). Looking at `lib/curl_setup.h`, it would have to be detected, alongside some "magic" conditions for including it. Then they have slightly different signatures, needing casts.

Sounds overkill for a standalone example. I think excluding it is a reasonable solution.

**Comment by jay** [comment]
I disabled it for Amiga by defining an alternate main in the example for `__AMIGA__` that says amiga is not supported.

I had considered instead removing block_ip from the check_PROGRAMS list used by CMakeLists.txt and Makefile.am, but in Makefile.am the code would need to be POSIX compliant and there's no good way to do it without string comparison and something like a GNU automake text function [filter-out](https://www.gnu.org/savannah-checkouts/gnu/make/manual/html_node/Text-Functions.html#index-filter_002dout). Though technically there are some workarounds like use the shell to filter the variable and then re-import it, it would be messy.

File: docs/examples/block_ip.c [hunk 1]
```diff
@@ -29,6 +29,11 @@
  * filter IP addresses.
  */
 
+#ifdef __AMIGA__
+#include <stdio.h>
+int main(void) { printf("amigaos isn't supported.\n"); return 1; }
```
Comment (vszakats) [inline docs/examples/block_ip.c:34]:
```suggestion
int main(void) { printf("AmigaOS is not supported.\n"); return 1; }
```
Tiny nit to sync casing with other mentions (and drop the contraction while there).

**Review by vszakats — APPROVED** [review]

**Comment by vszakats** [comment]
> I disabled it for Amiga by defining an alternate main in the example for `__AMIGA__` that says amiga is not supported.
> 
> I had considered instead removing block_ip, when amiga host, from the check_PROGRAMS list after it's imported by CMakeLists.txt and Makefile.am. In cmake it would be easy `list(REMOVE_ITEM check_PROGRAMS "block_ip")`, however in Makefile.am the code would need to be POSIX compliant and there's no good way to do it without string comparison ($(host_os) = amigaos) and something like a GNU automake text function [filter-out](https://www.gnu.org/savannah-checkouts/gnu/make/manual/html_node/Text-Functions.html#index-filter_002dout). Though technically there are some workarounds like use the shell to filter the variable and then re-import it, it would be messy. Also there's POSIX substitution but it only works with suffixes, eg ` $(check_PROGRAMS:block_ip=)`

Agreed, it sounds messy on the build-level.

I think it's perfectly fine to handle it in C. Somebody may compile
it as-is as a C source (outside cmake or autotools), and this is now
handled too.

**Pull request closed without merging**